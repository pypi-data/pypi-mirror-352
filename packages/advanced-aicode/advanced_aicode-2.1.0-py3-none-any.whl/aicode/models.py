from dataclasses import dataclass

from aicode.args import Args

CHAT_GPT = "openai/gpt-4o"


@dataclass
class Model:
    name: str
    description: str
    model_str: str


MODELS = {
    "chatgpt": Model("gpt-4o", "The GPT-4o model.", CHAT_GPT),
    "o3": Model("o3", "The O3 model.", "openai/o3"),
    "claude": Model(
        "claude", "The Claude model.", "anthropic/claude-3-7-sonnet-latest"
    ),
    "deepseek": Model(
        "deepseek",
        "The deepseek model.",
        "deepseek",
    ),
    "gemini": Model(
        "gemini",
        "The Google Gemini model.",
        "gemini/gemini-2.5-pro-preview-05-0",
    ),
}

CLAUD3_MODELS = {MODELS["claude"].model_str}

MODEL_CHOICES = list(MODELS.keys())


def get_model(
    args: Args,
    anthropic_key: str | None,
    openai_key: str | None,
    gemini_key: str | None,
) -> str:

    deep = not args.no_architect
    if deep:
        if args.model is not None and args.model in MODELS:
            return MODELS[args.model].model_str
        if openai_key is not None:
            return "openai/gpt-4o"

    if args.claude:
        assert "claude" in MODELS
        return MODELS["claude"].model_str
    elif args.chatgpt:
        return CHAT_GPT
    elif args.model is not None:
        return MODELS[args.model].model_str if args.model in MODELS else args.model
    elif anthropic_key is not None:
        return MODELS["claude"].model_str
    elif openai_key is not None:
        return CHAT_GPT
    elif gemini_key is not None:
        return MODELS["gemini"].model_str
    return MODELS["claude"].model_str
