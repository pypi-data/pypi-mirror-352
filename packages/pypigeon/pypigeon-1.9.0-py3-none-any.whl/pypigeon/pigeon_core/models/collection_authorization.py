from typing import Any
from typing import cast
from typing import Dict
from typing import List
from typing import Type
from typing import TypeVar
from typing import Union

from attrs import define as _attrs_define

from ..models.collection_authorization_grant import CollectionAuthorizationGrant
from ..models.collection_authorization_inherited import CollectionAuthorizationInherited
from ..models.collection_authorization_permit import CollectionAuthorizationPermit
from ..types import UNSET
from ..types import Unset


T = TypeVar("T", bound="CollectionAuthorization")


@_attrs_define
class CollectionAuthorization:
    """CollectionAuthorization model

    Attributes:
        myself (CollectionAuthorizationGrant):
        owners (List[str]):
        public (CollectionAuthorizationGrant):
        inherited (Union[Unset, CollectionAuthorizationInherited]):
        permit (Union[Unset, CollectionAuthorizationPermit]):
    """

    myself: "CollectionAuthorizationGrant"
    owners: List[str]
    public: "CollectionAuthorizationGrant"
    inherited: Union[Unset, "CollectionAuthorizationInherited"] = UNSET
    permit: Union[Unset, "CollectionAuthorizationPermit"] = UNSET

    def to_dict(self) -> Dict[str, Any]:
        """Convert to a dict"""
        myself = self.myself.to_dict()
        owners = self.owners

        public = self.public.to_dict()
        inherited: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.inherited, Unset):
            inherited = self.inherited.to_dict()
        permit: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.permit, Unset):
            permit = self.permit.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(
            {
                "myself": myself,
                "owners": owners,
                "public": public,
            }
        )
        if inherited is not UNSET:
            field_dict["inherited"] = inherited
        if permit is not UNSET:
            field_dict["permit"] = permit

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        """Create an instance of :py:class:`CollectionAuthorization` from a dict"""
        d = src_dict.copy()
        myself = CollectionAuthorizationGrant.from_dict(d.pop("myself"))

        owners = cast(List[str], d.pop("owners"))

        public = CollectionAuthorizationGrant.from_dict(d.pop("public"))

        _inherited = d.pop("inherited", UNSET)
        inherited: Union[Unset, CollectionAuthorizationInherited]
        if isinstance(_inherited, Unset):
            inherited = UNSET
        else:
            inherited = CollectionAuthorizationInherited.from_dict(_inherited)

        _permit = d.pop("permit", UNSET)
        permit: Union[Unset, CollectionAuthorizationPermit]
        if isinstance(_permit, Unset):
            permit = UNSET
        else:
            permit = CollectionAuthorizationPermit.from_dict(_permit)

        collection_authorization = cls(
            myself=myself,
            owners=owners,
            public=public,
            inherited=inherited,
            permit=permit,
        )

        return collection_authorization
