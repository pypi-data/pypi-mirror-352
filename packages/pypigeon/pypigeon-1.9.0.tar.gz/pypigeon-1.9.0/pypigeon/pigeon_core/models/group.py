import datetime
from typing import Any
from typing import Dict
from typing import Type
from typing import TypeVar

from attrs import define as _attrs_define
from dateutil.parser import isoparse

from ..models.group_metadata import GroupMetadata
from ..models.visibility_level import VisibilityLevel


T = TypeVar("T", bound="Group")


@_attrs_define
class Group:
    """Group model

    Attributes:
        created_on (datetime.datetime):
        id (str):
        metadata (GroupMetadata):
        name (str):
        visibility_level (VisibilityLevel):
    """

    created_on: datetime.datetime
    id: str
    metadata: "GroupMetadata"
    name: str
    visibility_level: VisibilityLevel

    def to_dict(self) -> Dict[str, Any]:
        """Convert to a dict"""
        created_on = self.created_on.isoformat()
        id = self.id
        metadata = self.metadata.to_dict()
        name = self.name
        visibility_level = self.visibility_level.value

        field_dict: Dict[str, Any] = {}
        field_dict.update(
            {
                "createdOn": created_on,
                "id": id,
                "metadata": metadata,
                "name": name,
                "visibilityLevel": visibility_level,
            }
        )

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        """Create an instance of :py:class:`Group` from a dict"""
        d = src_dict.copy()
        created_on = isoparse(d.pop("createdOn"))

        id = d.pop("id")

        metadata = GroupMetadata.from_dict(d.pop("metadata"))

        name = d.pop("name")

        visibility_level = VisibilityLevel(d.pop("visibilityLevel"))

        group = cls(
            created_on=created_on,
            id=id,
            metadata=metadata,
            name=name,
            visibility_level=visibility_level,
        )

        return group
