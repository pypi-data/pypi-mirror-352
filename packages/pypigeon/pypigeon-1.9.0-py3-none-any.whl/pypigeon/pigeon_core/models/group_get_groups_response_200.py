from typing import Any
from typing import Dict
from typing import List
from typing import Type
from typing import TypeVar

from attrs import define as _attrs_define

from ..models.group import Group
from ..models.pagination import Pagination


T = TypeVar("T", bound="GroupGetGroupsResponse200")


@_attrs_define
class GroupGetGroupsResponse200:
    """GroupGetGroupsResponse200 model

    Attributes:
        groups (List['Group']):
        pagination (Pagination):
    """

    groups: List["Group"]
    pagination: "Pagination"

    def to_dict(self) -> Dict[str, Any]:
        """Convert to a dict"""
        groups = []
        for groups_item_data in self.groups:
            groups_item = groups_item_data.to_dict()
            groups.append(groups_item)

        pagination = self.pagination.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(
            {
                "groups": groups,
                "pagination": pagination,
            }
        )

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        """Create an instance of :py:class:`GroupGetGroupsResponse200` from a dict"""
        d = src_dict.copy()
        groups = []
        _groups = d.pop("groups")
        for groups_item_data in _groups:
            groups_item = Group.from_dict(groups_item_data)

            groups.append(groups_item)

        pagination = Pagination.from_dict(d.pop("pagination"))

        group_get_groups_response_200 = cls(
            groups=groups,
            pagination=pagination,
        )

        return group_get_groups_response_200
