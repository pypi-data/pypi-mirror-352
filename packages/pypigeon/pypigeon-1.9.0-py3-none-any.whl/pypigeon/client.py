"""Login and general Pigeon operations"""
from collections.abc import Iterator
from urllib.parse import urlunparse

from .collections import PigeonCollection
from .pigeon_core import AuthenticatedClient as CoreAuthClient
from .pigeon_core import Client as CoreClient
from .pigeon_core import Login
from .pigeon_core import Paginator
from .pigeon_core.api.auth import auth_new_session
from .pigeon_core.api.collections import collections_create_collection
from .pigeon_core.api.collections import collections_get_collections
from .pigeon_core.models import CollectionsGetCollectionsCollectionList
from .pigeon_core.models import NewCollection
from .pigeon_core.models import NewCollectionMetadata
from .pigeon_core.types import UNSET


class PigeonClient:
    """Main class for interacting with a Pigeon instance."""

    def __init__(self, core: CoreAuthClient) -> None:
        self._client = core

    def __del__(self) -> None:
        if self._client._client:
            self._client._client.close()

    def get_collection_by_name(self, collection_name: str) -> PigeonCollection:
        """Retrieve a single collection given its name.

        Raises:
            ValueError: if the collection could not be found or if more than one
                collection was found with the given name
        """
        collections = collections_get_collections.sync(
            client=self._client, name=collection_name
        )
        if len(collections.collections) == 1:
            return PigeonCollection(collections.collections[0], self._client)

        raise ValueError(f"could not find collection: {collection_name}")

    def all_collections(self) -> Iterator[PigeonCollection]:
        """Iterate over all collections visible to the user."""
        paginator = Paginator[CollectionsGetCollectionsCollectionList](
            collections_get_collections,
            self._client,
        )
        for page in paginator.paginate():
            for collection in page.collections:
                yield PigeonCollection(collection, self._client)

    def new_collection(
        self, name: str, description: str, metadata: dict[str, str] = {}
    ) -> PigeonCollection:
        """Create a new collection.

        Args:
            name: The short name of the collection
            description: A longer plain text description of the collection
            metadata: Optional key-value pairs of metadata on the collection

        """
        collection = collections_create_collection.sync(
            client=self._client,
            body=NewCollection(
                name=name,
                description=description,
                metadata=(
                    NewCollectionMetadata.from_dict(metadata) if metadata else UNSET
                ),
            ),
        )
        return PigeonCollection(collection, self._client)


def make_base_url(host: str, scheme: str = "https") -> str:
    """Create a base URL for a Pigeon instance.

    Args:
        host: The hostname of your Pigeon instance, such as 'pigeon.bioteam.net'.
        scheme: either 'https' or 'http'

    Returns:
        the Pigeon API base URL

    """
    return urlunparse([scheme, host, "/api/v1", None, None, None])


def login(host: str, scheme: str = "https") -> PigeonClient:
    """Login to a Pigeon instance.

    This interactive login method will print a URL that you will need
    to visit in order to activate the login session. The function will
    wait to return until the session has successfully been activated.

    Args:
        host: The hostname of your Pigeon instance, such as 'pigeon.bioteam.net'.
        scheme: either 'https' or 'http'

    Returns:
        A :py:class:`PigeonClient`

    """

    # get a session ID
    core = CoreClient(base_url=make_base_url(host, scheme))
    session_token = auth_new_session.sync(client=core)

    # construct the URL to activate the session
    activation_url = urlunparse(
        [
            scheme,
            host,
            f"/login/activate/{session_token.activation_code}",
            None,
            None,
            None,
        ]
    )

    print("To activate your session, visit the URL below:")
    print("  ", activation_url)
    print()
    print("Waiting for session activation...")

    # wait for the session to be activated
    authenticated = Login(core).with_session_id(session_token.session_id, wait=True)

    return PigeonClient(authenticated)
