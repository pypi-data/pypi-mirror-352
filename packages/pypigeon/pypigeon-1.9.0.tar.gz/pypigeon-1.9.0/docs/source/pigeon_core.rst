Pigeon Core library
===================

The :py:mod:`pypigeon.pigeon_core` module is a set of low-level
endpoints and data models that have been automatically generated from
Pigeon's OpenAPI spec. For every Pigeon API endpoint, there is a
single API method in :py:mod:`pypigeon.pigeon_core.api`, and for every
request and response body, there is a corresponding data model defined
in :py:mod:`pypigeon.pigeon_core.models`.

Since this module is automatically generated, any missing or
insubstantial documentation should be addressed by adding the
appropriate fields to the main ``pigeon-api.yaml`` file. See the
:ref:`Autogeneration` section for more information.

Usage
-----

First, create a client::

    from pypigeon.pigeon_core import Client

    client = Client(base_url="https://pigeon.bioteam.net/api/v1")

If the endpoints you're going to hit require authentication, login
using the :py:func:`pypigeon.pigeon_core.login.Login` helper::

    from pypigeon.pigeon_core import Client
    from pypigeon.pigeon_core import Login

    client = Login(
        Client(base_url="https://pigeon.bioteam.net/api/v1")
    ).with_password('myuser', 'mypassword')

Now call your endpoint and use your models::

    from pypigeon.pigeon_core.models import CollectionsGetCollectionsCollectionList as CollectionList
    from pypigeon.pigeon_core.api.collections import collections_get_collections
    from pypigeon.pigeon_core.types import Response

    # the context manager is recommended in order to clean up any HTTP connection pool,
    # see https://www.python-httpx.org/advanced/
    with client as client:
        coll_list: CollectionList = collections_get_collections.sync(client=client)

        # or if you need more info (e.g. status_code)
        response: Response[CollectionList] = collections_get_collections.sync_detailed(client=client)

Or do the same thing with an async version::

    from pypigeon.pigeon_core.models import CollectionsGetCollectionsCollectionList as CollectionList
    from pypigeon.pigeon_core.api.collections import collections_get_collections
    from pypigeon.pigeon_core.types import Response

    async with client as client:
        my_data: CollectionList = await collections_get_collections.asyncio(client=client)
        response: Response[CollectionList] = await collections_get_collections.asyncio_detailed(client=client)

By default, when you're calling an HTTPS API it will attempt to verify
that SSL is working correctly. Using certificate verification is
highly recommended most of the time, but sometimes you may need to
authenticate to a server (especially an internal server) using a
custom certificate bundle::

    client = AuthenticatedClient(
        base_url="https://pigeon.local/api/v1",
        token="SuperSecretToken",
        verify_ssl="/path/to/certificate_bundle.pem",
    )

You can also disable certificate validation altogether, but beware
that **this is a security risk**::

    client = AuthenticatedClient(
        base_url="https://pigeon.local/api/v1",
        token="SuperSecretToken",
        verify_ssl=False
    )

Things to know
--------------

1. Every path/method combo becomes a Python module with four functions:

    1. ``sync``: Blocking request that returns parsed data, if
       successful, or raises an exception otherwise.
    2. ``sync_detailed``: Blocking request that always returns a
       ``Request``, optionally with ``parsed`` set if the request was
       successful.
    3. ``asyncio``: Like ``sync`` but async instead of blocking
    4. ``asyncio_detailed``: Like ``sync_detailed`` but async instead of blocking

2. All path/query params, and bodies become method arguments.
3. If your endpoint had any tags on it, the first tag will be used as
   a module name for the function ("collections" in the above examples)
4. Any endpoint which did not have a tag will be in
   :py:mod:`pypigeon.pigeon_core.api.default`

Advanced customizations
-----------------------

There are more settings on the generated ``Client`` class which let you
control more runtime behavior, check out the docstring on that class
for more info. You can also customize the underlying ``httpx.Client`` or
``httpx.AsyncClient`` (depending on your use-case)::

    from pypigeon.pigeon_core import Client

    def log_request(request):
        print(f"Request event hook: {request.method} {request.url} - Waiting for response")

    def log_response(response):
        request = response.request
        print(f"Response event hook: {request.method} {request.url} - Status {response.status_code}")

    client = Client(
        base_url="https://pigeon.bioteam.net/api/v1",
        httpx_args={"event_hooks": {"request": [log_request], "response": [log_response]}},
    )

    # Or get the underlying httpx client to modify directly with client.get_httpx_client()
    # or client.get_async_httpx_client()

You can even set the httpx client directly, but beware that this will
override any existing settings (e.g., base_url)::

    import httpx
    from pypigeon.pigeon_core import Client

    client = Client(
        base_url="https://pigeon.bioteam.net/api/v1",
    )
    # Note that base_url needs to be re-set, as would any shared cookies, headers, etc.
    client.set_httpx_client(
        httpx.Client(
            base_url="https://pigeon.bioteam.net/api/v1", proxies="http://localhost:8030"
        )
    )

.. _Autogeneration:

Autogeneration
--------------

The core client code can be regenerated by running ``autogenerate.sh``
in the ``clients/pypigeon`` directory. This is automatically run by
the pre-commit hook system any time that ``pigeon-api.yaml`` is
edited, so the client code should always remain in sync with the API
specification.

The ``autogenerate.sh`` script contains several templates and
workarounds for various bugs and strange behavior in
openapi-python-client.

Documentation
-------------

All documentation can be found under the
:py:mod:`pypigeon.pigeon_core` API documentation section.

.. autosummary::
   :toctree: generated
   :template: custom-module-template.rst
   :recursive:

   pypigeon.pigeon_core.api
   pypigeon.pigeon_core.client
   pypigeon.pigeon_core.errors
   pypigeon.pigeon_core.login
   pypigeon.pigeon_core.models
   pypigeon.pigeon_core.paginator
   pypigeon.pigeon_core.types


Updating Documentation
^^^^^^^^^^^^^^^^^^^^^^

Since the documentation is autogenerated from the code, and the code
is autogenerated from the original ``pigeon-api.yaml`` API spec, any
updates to the :py:mod:`~pypigeon.pigeon_core` documentation should be
made in ``pigeon-api.yaml``. The mapping between fields in the source
OpenAPI document and the outputs in the generated code are:

* ``title:`` is essentially a rename for a given schema or endpoint.
  For example, if a schema is defined with the key
  ``dictionary_search_options:`` it will typically be autogenerated
  with the model title **DictionarySearchOptions**, but by providing
  ``title: Search Options for Dictionaries`` the model title will now
  be **SearchOptionsForDictionaries**. This should be used sparingly
  as it makes it harder to map back from the generated code to the
  original API definition.

* ``summary:`` is a short, one-line summary of a schema or endpoint.

* ``description:`` is a longer, multi-line description of the function
  of a schema or endpoint.
