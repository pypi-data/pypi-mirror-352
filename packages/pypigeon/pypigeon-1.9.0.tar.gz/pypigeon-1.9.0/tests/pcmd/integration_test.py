import argparse
import ast
import os
import random
import re
import shlex
from pathlib import Path
from urllib.request import urlopen

import pytest
from pypigeon.client import make_base_url
from pypigeon.pcmd import cli
from pypigeon.pcmd.commands import Commands as PcmdCommands
from pypigeon.pigeon_core import Client as CoreClient
from pypigeon.pigeon_core.api.user import user_create_user
from pypigeon.pigeon_core.login import Login
from pypigeon.pigeon_core.login import LoginError
from pypigeon.pigeon_core.login import to_token_file
from pypigeon.pigeon_core.models import NewUser
from yaml import dump as yaml_dump
from yaml import safe_load


# The tests in this test suite are designed to be run against a local
# development instance of Pigeon hosted at localhost:8170, similar to
# the Cypress tests in `pigeon-nuxt`. In order to be able to run these
# tests, be sure to have the local development instance running in the
# background with the command:
#
#    flask run --debug
#


@pytest.fixture
def local_user(tmp_path):
    # try to log in with the credentials
    username, password = "pcmd.tester", "insecure"
    base_url = make_base_url("localhost:8170", "http")
    try:
        authenticated = Login(CoreClient(base_url=base_url)).with_password(
            username, password
        )
    except LoginError:
        print("Login failed, trying to create new user...")
        user_create_user.sync(
            body=NewUser(
                given_name="pcmd",
                last_name="tester",
                user_email="john.doe@example.com",
                user_name=username,
                password=password,
            ),
            client=CoreClient(base_url=base_url),
        )
        authenticated = Login(CoreClient(base_url=base_url)).with_password(
            username, password
        )

    token_file = tmp_path / "token"
    to_token_file(authenticated, token_file)

    yield str(token_file)


def random_id():
    return f"rr{random.randint(0, 1048576):05x}rr"


def mask(rv):
    # mask UUIDs
    rv = re.sub(
        r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}",
        "00000000-0000-0000-0000-000000000000",
        rv,
    )
    # mask timestamps
    rv = re.sub(
        r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.[0-9+:]*)?",
        "1970-01-02T03:04:05.000000",
        rv,
    )
    # mask task worker
    rv = re.sub(r"worker: .*", "worker: pcmd-test-mask", rv)
    # mask random ID
    rv = re.sub(r"rr[0-9a-f]{5}rr", "random-id", rv)
    # mask bundle ID
    rv = re.sub(r"b-[a-zA-Z0-9]{8}", "b-00000000", rv)

    return rv


@pytest.fixture
def pcmd(local_user, capsys):
    def _cmd(args, masked=True):
        if isinstance(args, str):
            args = shlex.split(args)
        cli.main(["-H", "localhost:8170", "--insecure", "-k", local_user] + args)

        outv = capsys.readouterr()
        if not outv.err:
            rv = outv.out
            if masked:
                rv = mask(rv)
            return rv

        return outv

    yield _cmd


@pytest.fixture
def test_target_user(pcmd):
    response = user_create_user.sync(
        body=NewUser(
            given_name="pcmd",
            last_name="test target",
            user_email="john.doe@example.com",
            user_name="pcmd.test.target",
            password=random_id(),
        ),
        client=CoreClient(base_url=make_base_url("localhost:8170", "http")),
    )

    yield response.subject_id

    pcmd("users delete pcmd.test.target")


@pytest.fixture
def scratch_collection(pcmd):
    collection_name = f"pcmd-test.{random_id()}"
    rv = safe_load(pcmd(f"collection new {collection_name}", masked=False))

    yield collection_name, rv["id"]

    pcmd(f"collection delete {rv['id']}")


def test_pcmd_auth_whoami(pcmd, snapshot):
    pcmd('auth set-metadata x-tos \'{"toc":3,"coc":6,"priv":3,"com":1}\'')

    assert pcmd("auth whoami") == snapshot


# NOTE: If the following tests are failing with "Unexpected status
# code: 403", you should grant the `pcmd.tester` user full admin
# access to your local development instance with the command:
#
#   flask admin grant pcmd.tester
#


def test_pcmd_tasks(pcmd, snapshot):
    stats = safe_load(pcmd("tasks stats"))
    assert stats.keys() == {"queued", "results", "schedules"}

    schedules = safe_load(pcmd("tasks schedules"))["schedules"]
    assert len(schedules) >= 2

    results = safe_load(pcmd("tasks results --idemkey tasks-backends-sqlite-vacuum"))[
        "results"
    ]
    if len(results) > 0:
        assert results[0] == snapshot


def test_pcmd_users(pcmd, test_target_user):
    rv = pcmd("users list")
    assert set(rv.splitlines()[0].split()) == {
        "account_id",
        "created_on",
        "metadata",
        "subject_id",
        "user_name",
        "visibility_level",
        "given_name",
        "groups",
        "last_name",
        "user_email",
    }
    assert "pcmd.test.target" in rv

    # disable and enable
    rv = safe_load(pcmd("users disable pcmd.test.target"))
    assert rv["disabled_on"] is not None

    rv = safe_load(pcmd("users enable pcmd.test.target"))
    assert rv["disabled_on"] is None


def test_pcmd_groups(pcmd, test_target_user, snapshot):
    group_name = f"test-pcmd-group.{random_id()}"
    new_group = pcmd(f"groups new {group_name}", masked=False)
    group_id = safe_load(new_group)["id"]
    assert mask(new_group) == snapshot

    groups = safe_load(pcmd("groups list"))["groups"]
    assert safe_load(mask(new_group)) in groups

    pcmd(f"groups join {group_id} {test_target_user}")

    assert len(safe_load(pcmd(f"groups list-members {group_id}"))["members"]) == 2

    pcmd(f"groups update --name new-test-group {group_id}")

    pcmd(f"groups delete {group_id}")


def test_pcmd_accounts(pcmd, snapshot, tmp_path, scratch_collection):
    collection_name, collection_id = scratch_collection

    account_name = f"test-pcmd-account.{random_id()}"
    new_account = pcmd(f"accounts new {account_name} --description heyyo", masked=False)
    account_id = safe_load(new_account)["id"]
    assert mask(new_account) == snapshot

    accounts = safe_load(pcmd("accounts list", masked=False))["accounts"]
    assert safe_load(new_account) in accounts

    # update space
    index_file = tmp_path / "index.md"
    with open(index_file, "w") as fp:
        fp.write("# Hello world!\n")
    pcmd(f"collection upload {collection_name} {index_file}")
    pcmd(f"accounts update-space {account_id} {collection_id}")
    response = urlopen(
        f"http://localhost:8170/api/v1/account/{account_id}/pages/index.md"
    )
    assert response.code == 200

    pcmd(f"accounts delete {account_id}")

    accounts = safe_load(pcmd("accounts list", masked=False))["accounts"]
    assert safe_load(new_account) not in accounts


def test_pcmd_admin_grant(pcmd, test_target_user):
    pcmd(f"admin grant-admin -o group {test_target_user}")

    rv = pcmd("admin list-admins", masked=False)
    assert test_target_user in rv

    pcmd(f"admin revoke-admin {test_target_user}")

    rv = pcmd("admin list-admins", masked=False)
    assert test_target_user not in rv


def test_pcmd_admin_config(pcmd, tmp_path, capsys):
    rv = safe_load(pcmd("admin config get"))
    assert "datastores" in rv
    assert "authentication" in rv

    pcmd(
        "admin config set datastores.pcmd_test"
        + f' \'{{"enabled": false, "path": "{tmp_path}", "type": "local"}}\''
    )

    rv = safe_load(pcmd("admin config get", masked=False))
    assert "pcmd_test" in rv["datastores"]

    newdir = tmp_path / "new"
    rv["datastores"]["pcmd_test"]["path"] = str(newdir)

    config_fn = tmp_path / "config.yaml"
    with open(config_fn, "w") as fp:
        yaml_dump(rv, fp)

    pcmd(f"admin config put {config_fn}")

    rv = safe_load(pcmd("admin config get"))
    assert "pcmd_test" in rv["datastores"]
    assert rv["datastores"]["pcmd_test"]["path"] == str(newdir)

    pcmd("admin config delete datastores.pcmd_test")

    rv = safe_load(pcmd("admin config get"))
    assert "pcmd_test" not in rv["datastores"]


def test_pcmd_collection(pcmd, tmp_path, scratch_collection):
    collection_name, collection_id = scratch_collection
    rv = pcmd("collection list", masked=False)
    assert collection_name in rv
    assert collection_id in rv

    test_file = tmp_path / "test.txt"
    with open(test_file, "w") as fp:
        fp.write("hello world this is a test\n")

    pcmd(f"collection upload {collection_name} {test_file}")

    test_file.unlink()

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    pcmd(f"collection download {collection_name} test.txt")
    os.chdir(old_cwd)

    assert test_file.exists()


def test_pcmd_pdd(pcmd, scratch_collection, snapshot):
    collection_name, collection_id = scratch_collection
    common_pdd_fn = Path(__file__).parent / "common_pdd.json"

    rv = pcmd(f"pdd upload {collection_id} {common_pdd_fn}", masked=False)
    assert mask(rv) == snapshot
    safe_load(rv)["id"]

    assert pcmd(f"pdd list {collection_name} common_pdd.json") == snapshot

    pcmd(f"pdd set-order {collection_name} common_pdd.json 6fr70ICV O8zPEsiM")

    assert pcmd(f"pdd list {collection_name} common_pdd.json") == snapshot

    assert (
        pcmd(f"pdd bundle {collection_name} common_pdd.json MyBundle O8zPEsiM 6fr70ICV")
        == snapshot
    )

    assert pcmd(f"pdd list {collection_name} common_pdd.json") == snapshot


def test_pcmd_cdeset(pcmd, scratch_collection):
    collection_name, collection_id = scratch_collection
    common_pdd_fn = Path(__file__).parent / "common_pdd.json"

    rv = pcmd(f"pdd upload {collection_id} {common_pdd_fn}", masked=False)
    item_id = safe_load(rv)["id"]

    cdeset_name = f"pcmd.{random_id()}"
    pcmd(f"cdeset new {cdeset_name} 'pcmd test cde set'")

    rv = pcmd("cdeset list", masked=False)
    assert cdeset_name in rv

    pcmd(f"cdeset copy-from-pdd {cdeset_name} {collection_id} LIVE {item_id}")

    rv = safe_load(pcmd("cdeset list", masked=False))
    cdeset = {"f": c for c in rv["cdesets"] if c["name"] == cdeset_name}.get("f")
    assert cdeset["num_cdes"] == 2
    assert cdeset["description"] == "pcmd test cde set"

    pcmd(f"cdeset delete {cdeset_name}")


###
### Coverage Verification
###


class CoverageScanner(ast.NodeVisitor):
    def __init__(self):
        super().__init__()
        self.actions = {}

    def register_actions(self, parser, actlevel=None):
        if actlevel is None:
            actlevel = self.actions

        actions = parser._get_positional_actions()
        if actions and actions[0].choices:
            for arg, subparser in actions[0].choices.items():
                self.register_actions(subparser, actlevel.setdefault(arg, {}))

    def coverage_needed(self):
        needed = set()

        def _walk(d, s=None):
            if s is None:
                s = []
            if "SEEN" in d:
                return
            if not d:
                needed.add(" ".join(s))
            for k in d:
                _walk(d[k], s + [k])

        _walk(self.actions)
        return needed

    def visit_Call(self, node):
        if not isinstance(node.func, ast.Name):
            return self.generic_visit(node)
        if node.func.id != "pcmd":
            return self.generic_visit(node)
        if not node.args:
            return self.generic_visit(node)

        def _resolve(arg):
            if isinstance(arg, ast.Constant):
                return arg.value
            elif isinstance(arg, ast.JoinedStr):
                return _resolve(arg.values[0])
            elif isinstance(arg, ast.BinOp):
                return _resolve(arg.left)
            else:
                print(ast.dump(node, indent=2))
                raise Exception(
                    f"unexpected pcmd call on line {node.lineno}: {ast.unparse(node)}"
                )

        pcmd_arg = _resolve(node.args[0])

        actlevel = self.actions
        for word in shlex.split(pcmd_arg):
            if word in actlevel:
                actlevel = actlevel[word]

        if not actlevel:
            actlevel["SEEN"] = True


def test_pcmd_coverage():
    # This checks all of the pcmd() calls in the AST of this file and
    # matches them up with commands defined in the pcmd command set.
    # Any commands not called will show up in
    # `scanner.coverage_needed()`.

    argparser = argparse.ArgumentParser()
    PcmdCommands._add_to_parser(
        argparser.add_subparsers(title="commands", required=True)
    )
    scanner = CoverageScanner()
    scanner.register_actions(argparser)

    test_ast = ast.parse(open(__file__).read(), filename=__file__)
    scanner.visit(test_ast)

    assert scanner.coverage_needed() == set()
