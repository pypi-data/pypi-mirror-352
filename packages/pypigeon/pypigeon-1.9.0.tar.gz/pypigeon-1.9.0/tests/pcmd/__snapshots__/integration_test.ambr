# serializer version: 1
# name: test_pcmd_accounts
  '''
  createdOn: '1970-01-02T03:04:05.000000'
  id: a:00000000-0000-0000-0000-000000000000
  isPersonal: false
  membersGroupId: g:00000000-0000-0000-0000-000000000000
  metadata:
    x-description: heyyo
  name: test-pcmd-account.random-id
  visibilityLevel: public
  
  
  '''
# ---
# name: test_pcmd_auth_whoami
  '''
  expires: '1970-01-02T03:04:05.000000'
  user:
    account_id: a:00000000-0000-0000-0000-000000000000
    email: john.doe@example.com
    has_operations:
    - account
    - admin.cde
    - admin.config
    - admin.grantadmin
    - admin.metadata
    - admin.metrics
    - admin.tasks
    - admin.users
    - group
    identity:
      fqda: pcmd.tester@localhost
      is_local_user: true
      primary_id: 00000000-0000-0000-0000-000000000000
      user_name: pcmd.tester
    image: null
    metadata:
      x-tos: '{"toc":3,"coc":6,"priv":3,"com":1}'
    name: pcmd tester
  
  
  '''
# ---
# name: test_pcmd_groups
  '''
  createdOn: '1970-01-02T03:04:05.000000'
  id: g:00000000-0000-0000-0000-000000000000
  metadata: {}
  name: test-pcmd-group.random-id
  visibilityLevel: unlisted
  
  
  '''
# ---
# name: test_pcmd_pdd
  '''
  id: 00000000-0000-0000-0000-000000000000
  
  '''
# ---
# name: test_pcmd_pdd.1
  '''
  [O8zPEsiM] Age
  [6fr70ICV] Zip Code
  
  '''
# ---
# name: test_pcmd_pdd.2
  '''
  [6fr70ICV] Zip Code
  [O8zPEsiM] Age
  
  '''
# ---
# name: test_pcmd_pdd.3
  '''
  New bundle name: MyBundle
  Element IDs: ['O8zPEsiM', '6fr70ICV']
  
  New bundle ID: b-00000000
  
  '''
# ---
# name: test_pcmd_pdd.4
  '''
  MyBundle
     [O8zPEsiM] Age
     [6fr70ICV] Zip Code
  
  '''
# ---
# name: test_pcmd_tasks
  dict({
    'data': None,
    'errors': list([
    ]),
    'event_time': '1970-01-02T03:04:05.000000',
    'id': '00000000-0000-0000-0000-000000000000',
    'start_time': '1970-01-02T03:04:05.000000',
    'status': 'OK',
    'task_id': '00000000-0000-0000-0000-000000000000',
    'task_idemkey': 'tasks-backends-sqlite-vacuum',
    'worker': 'pcmd-test-mask',
  })
# ---
